File: backend/src/routes/progress.ts
import { Router } from 'express';
import { query } from '../database/db.js';
import { authenticate, AuthRequest } from '../middleware/auth.js';
import type { ProgressCreate, ProgressUpdate } from '../types/index.js';

const router = Router();

// Add progress
router.post('/', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { category, metric, value, unit, notes, date } = req.body as ProgressCreate;

    if (!category || !metric || value === undefined || !date) {
      return res.status(400).json({ error: 'Category, metric, value, and date are required' });
    }

    const insertResult = await query(
      `INSERT INTO progress (user_id, category, metric, value, unit, notes, date)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [req.userId, category, metric, value, unit || null, notes || null, date]
    );

    // Fetch the inserted progress
    const result = await query(
      'SELECT * FROM progress WHERE id = ?',
      [insertResult.rows[0].lastInsertRowid]
    );

    res.status(201).json(result.rows[0]);
  } catch (error) {
    next(error);
  }
});

// Get user progress
router.get('/', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { limit = '100', offset = '0', category, startDate, endDate } = req.query;

    let queryText = 'SELECT * FROM progress WHERE user_id = ?';
    const values: any[] = [req.userId];

    if (category) {
      queryText += ` AND category = ?`;
      values.push(category);
    }

    if (startDate) {
      queryText += ` AND date >= ?`;
      values.push(startDate);
    }

    if (endDate) {
      queryText += ` AND date <= ?`;
      values.push(endDate);
    }

    queryText += ` ORDER BY date DESC LIMIT ? OFFSET ?`;
    values.push(limit, offset);

    const result = await query(queryText, values);

    res.json(result.rows);
  } catch (error) {
    next(error);
  }
});

// Get progress by ID
router.get('/:progressId', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const progressId = parseInt(req.params.progressId);

    const result = await query(
      'SELECT * FROM progress WHERE id = ? AND user_id = ?',
      [progressId, req.userId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Progress not found' });
    }

    res.json(result.rows[0]);
  } catch (error) {
    next(error);
  }
});

// Update progress
router.put('/:progressId', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const progressId = parseInt(req.params.progressId);
    const { category, metric, value, unit, notes, date } = req.body as ProgressUpdate;

    // Check if progress exists and belongs to user
    const existingResult = await query(
      'SELECT * FROM progress WHERE id = ? AND user_id = ?',
      [progressId, req.userId]
    );

    if (existingResult.rows.length === 0) {
      return res.status(404).json({ error: 'Progress not found' });
    }

    const updates: string[] = [];
    const values: any[] = [];

    if (category !== undefined) {
      updates.push(`category = ?`);
      values.push(category);
    }
    if (metric !== undefined) {
      updates.push(`metric = ?`);
      values.push(metric);
    }
    if (value !== undefined) {
      updates.push(`value = ?`);
      values.push(value);
    }
    if (unit !== undefined) {
      updates.push(`unit = ?`);
      values.push(unit);
    }
    if (notes !== undefined) {
      updates.push(`notes = ?`);
      values.push(notes);
    }
    if (date !== undefined) {
      updates.push(`date = ?`);
      values.push(date);
    }

    if (updates.length === 0) {
      return res.status(400).json({ error: 'No fields to update' });
    }

    updates.push(`updated_at = datetime('now')`);
    values.push(progressId);

    await query(
      `UPDATE progress SET ${updates.join(', ')} WHERE id = ?`,
      values
    );

    // Fetch updated progress
    const fetchResult = await query(
      'SELECT * FROM progress WHERE id = ?',
      [progressId]
    );

    res.json(fetchResult.rows[0]);
  } catch (error) {
    next(error);
  }
});

// Delete progress
router.delete('/:progressId', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const progressId = parseInt(req.params.progressId);

    // Check if progress exists
    const existingResult = await query(
      'SELECT id FROM progress WHERE id = ? AND user_id = ?',
      [progressId, req.userId]
    );

    if (existingResult.rows.length === 0) {
      return res.status(404).json({ error: 'Progress not found' });
    }

    await query(
      'DELETE FROM progress WHERE id = ? AND user_id = ?',
      [progressId, req.userId]
    );

    res.json({ message: 'Progress deleted successfully' });
  } catch (error) {
    next(error);
  }
});

export default router;

File: src-tauri/src/commands/progress.rs
use tauri::State;
use sqlx::Row;
use crate::database::Database;
use crate::models::*;

// Progress commands
#[tauri::command]
pub async fn add_progress(
    db: State<'_, Database>,
    user_id: i64,
    progress_data: ProgressCreate,
) -> Result<Progress, String> {
    let result = sqlx::query(
        "INSERT INTO progress (user_id, category, metric, value, unit, notes, date) VALUES (?, ?, ?, ?, ?, ?, ?)"
    )
    .bind(user_id)
    .bind(&progress_data.category)
    .bind(&progress_data.metric)
    .bind(progress_data.value)
    .bind(&progress_data.unit)
    .bind(&progress_data.notes)
    .bind(progress_data.date)
    .execute(db.get_pool())
    .await
    .map_err(|e| e.to_string())?;

    let progress_id = result.last_insert_rowid();

    let progress_row = sqlx::query("SELECT * FROM progress WHERE id = ?")
        .bind(progress_id)
        .fetch_one(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    Ok(Progress {
        id: progress_row.get("id"),
        user_id: progress_row.get("user_id"),
        category: progress_row.get("category"),
        metric: progress_row.get("metric"),
        value: progress_row.get("value"),
        unit: progress_row.get("unit"),
        notes: progress_row.get("notes"),
        date: progress_row.get("date"),
        created_at: progress_row.get("created_at"),
        updated_at: progress_row.get("updated_at"),
    })
}

#[tauri::command]
pub async fn get_user_progress(
    db: State<'_, Database>,
    user_id: i64,
    limit: Option<i64>,
    offset: Option<i64>,
) -> Result<Vec<Progress>, String> {
    let limit = limit.unwrap_or(100);
    let offset = offset.unwrap_or(0);

    let rows = sqlx::query("SELECT * FROM progress WHERE user_id = ? ORDER BY date DESC, created_at DESC LIMIT ? OFFSET ?")
        .bind(user_id)
        .bind(limit)
        .bind(offset)
        .fetch_all(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    let progress_entries: Vec<Progress> = rows.into_iter().map(|row| Progress {
        id: row.get("id"),
        user_id: row.get("user_id"),
        category: row.get("category"),
        metric: row.get("metric"),
        value: row.get("value"),
        unit: row.get("unit"),
        notes: row.get("notes"),
        date: row.get("date"),
        created_at: row.get("created_at"),
        updated_at: row.get("updated_at"),
    }).collect();

    Ok(progress_entries)
}

#[tauri::command]
pub async fn get_user_progress_by_id(
    db: State<'_, Database>,
    target_user_id: i64,
    limit: Option<i64>,
    offset: Option<i64>,
) -> Result<Vec<Progress>, String> {
    let limit = limit.unwrap_or(100);
    let offset = offset.unwrap_or(0);

    let rows = sqlx::query("SELECT * FROM progress WHERE user_id = ? ORDER BY date DESC, created_at DESC LIMIT ? OFFSET ?")
        .bind(target_user_id)
        .bind(limit)
        .bind(offset)
        .fetch_all(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    let progress_entries: Vec<Progress> = rows.into_iter().map(|row| Progress {
        id: row.get("id"),
        user_id: row.get("user_id"),
        category: row.get("category"),
        metric: row.get("metric"),
        value: row.get("value"),
        unit: row.get("unit"),
        notes: row.get("notes"),
        date: row.get("date"),
        created_at: row.get("created_at"),
        updated_at: row.get("updated_at"),
    }).collect();

    Ok(progress_entries)
}

#[tauri::command]
pub async fn update_progress(
    db: State<'_, Database>,
    progress_id: i64,
    user_id: i64,
    update_data: ProgressUpdate,
) -> Result<Progress, String> {
    // Verify ownership
    let ownership_check = sqlx::query("SELECT id FROM progress WHERE id = ? AND user_id = ?")
        .bind(progress_id)
        .bind(user_id)
        .fetch_optional(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    if ownership_check.is_none() {
        return Err("Progress entry not found".to_string());
    }

    let mut has_updates = false;

    if let Some(category) = &update_data.category {
        sqlx::query("UPDATE progress SET category = ? WHERE id = ?")
            .bind(category)
            .bind(progress_id)
            .execute(db.get_pool())
            .await
            .map_err(|e| e.to_string())?;
        has_updates = true;
    }

    if let Some(metric) = &update_data.metric {
        sqlx::query("UPDATE progress SET metric = ? WHERE id = ?")
            .bind(metric)
            .bind(progress_id)
            .execute(db.get_pool())
            .await
            .map_err(|e| e.to_string())?;
        has_updates = true;
    }

    if let Some(value) = update_data.value {
        sqlx::query("UPDATE progress SET value = ? WHERE id = ?")
            .bind(value)
            .bind(progress_id)
            .execute(db.get_pool())
            .await
            .map_err(|e| e.to_string())?;
        has_updates = true;
    }

    if let Some(unit) = &update_data.unit {
        sqlx::query("UPDATE progress SET unit = ? WHERE id = ?")
            .bind(unit)
            .bind(progress_id)
            .execute(db.get_pool())
            .await
            .map_err(|e| e.to_string())?;
        has_updates = true;
    }

    if let Some(notes) = &update_data.notes {
        sqlx::query("UPDATE progress SET notes = ? WHERE id = ?")
            .bind(notes)
            .bind(progress_id)
            .execute(db.get_pool())
            .await
            .map_err(|e| e.to_string())?;
        has_updates = true;
    }

    if let Some(date) = update_data.date {
        sqlx::query("UPDATE progress SET date = ? WHERE id = ?")
            .bind(date)
            .bind(progress_id)
            .execute(db.get_pool())
            .await
            .map_err(|e| e.to_string())?;
        has_updates = true;
    }

    if !has_updates {
        return Err("No fields to update".to_string());
    }

    sqlx::query("UPDATE progress SET updated_at = datetime('now') WHERE id = ?")
        .bind(progress_id)
        .execute(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    // Fetch updated progress
    let progress_row = sqlx::query("SELECT * FROM progress WHERE id = ?")
        .bind(progress_id)
        .fetch_one(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    Ok(Progress {
        id: progress_row.get("id"),
        user_id: progress_row.get("user_id"),
        category: progress_row.get("category"),
        metric: progress_row.get("metric"),
        value: progress_row.get("value"),
        unit: progress_row.get("unit"),
        notes: progress_row.get("notes"),
        date: progress_row.get("date"),
        created_at: progress_row.get("created_at"),
        updated_at: progress_row.get("updated_at"),
    })
}

#[tauri::command]
pub async fn delete_progress(
    db: State<'_, Database>,
    progress_id: i64,
    user_id: i64,
) -> Result<(), String> {
    // Verify ownership
    let ownership_check = sqlx::query("SELECT id FROM progress WHERE id = ? AND user_id = ?")
        .bind(progress_id)
        .bind(user_id)
        .fetch_optional(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    if ownership_check.is_none() {
        return Err("Progress entry not found".to_string());
    }

    sqlx::query("DELETE FROM progress WHERE id = ?")
        .bind(progress_id)
        .execute(db.get_pool())
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}